File: C/team_service.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "team.h"


#define TRY( Func ) if ( (rc = (Func)) != SQLITE_OK ) return rc
#define INSERT_IF_UNIQUE( Func ) if ( (rc = (Func)) != SQLITE_OK  &&  rc != SQLITE_CONSTRAINT ) return rc


static int upsert_team_kicking_stats( sqlite3 *db, const team_kicking_stats_s *team_kicking_stats )
{
     int rc;

     if ( (rc = team_kicking_stats_t_create( db, team_kicking_stats )) == SQLITE_CONSTRAINT )
     {
          return team_kicking_stats_t_update( db, team_kicking_stats );
     }

     return rc;
}

static int save_team_kicking_stats( sqlite3 *db, const team_kicking_stats_s *team_kicking_stats )
{
     int rc;

     if ( team_kicking_stats == NULL ) return SQLITE_OK;

     for ( int i = 0; team_kicking_stats[i].team_id >= 0; ++i )
     {
          TRY( upsert_team_kicking_stats( db, &team_kicking_stats[i] ) );
     }

     return SQLITE_OK;
}

static int upsert_team_defense_stats( sqlite3 *db, const team_defense_stats_s *team_defense_stats )
{
     int rc;

     if ( (rc = team_defense_stats_t_create( db, team_defense_stats )) == SQLITE_CONSTRAINT )
     {
          return team_defense_stats_t_update( db, team_defense_stats );
     }

     return rc;
}

static int save_team_defense_stats( sqlite3 *db, const team_defense_stats_s *team_defense_stats )
{
     int rc;

     if ( team_defense_stats == NULL ) return SQLITE_OK;

     for ( int i = 0; team_defense_stats[i].team_id >= 0; ++i )
     {
          TRY( upsert_team_defense_stats( db, &team_defense_stats[i] ) );
     }

     return SQLITE_OK;
}

static int upsert_team_offense_stats( sqlite3 *db, const team_offense_stats_s *team_offense_stats )
{
     int rc;

     if ( (rc = team_offense_stats_t_create( db, team_offense_stats )) == SQLITE_CONSTRAINT )
     {
          return team_offense_stats_t_update( db, team_offense_stats );
     }

     return rc;
}

static int save_team_offense_stats( sqlite3 *db, const team_offense_stats_s *team_offense_stats )
{
     int rc;

     if ( team_offense_stats == NULL ) return SQLITE_OK;

     for ( int i = 0; team_offense_stats[i].team_id >= 0; ++i )
     {
          TRY( upsert_team_offense_stats( db, &team_offense_stats[i] ) );
     }

     return SQLITE_OK;
}

static int upsert_team_stats( sqlite3 *db, const team_stats_s *team_stats )
{
     int rc;

     if ( (rc = team_stats_t_create( db, team_stats )) == SQLITE_CONSTRAINT )
     {
          return team_stats_t_update( db, team_stats );
     }

     return rc;
}

static int save_team_stats( sqlite3 *db, const team_stats_s *team_stats )
{
     int rc;

     if ( team_stats == NULL ) return SQLITE_OK;

     for ( int i = 0; team_stats[i].team_id >= 0; ++i )
     {
          TRY( upsert_team_stats( db, &team_stats[i] ) );
     }

     return SQLITE_OK;
}

static int save_team_accolades( sqlite3 *db, const team_accolade_s *team_accolades )
{
     int rc;

     if ( team_accolades == NULL ) return SQLITE_OK;

     for ( int i = 0; team_accolades[i].team_id >= 0; ++i )
     {
          INSERT_IF_UNIQUE( team_accolades_t_create( db, &team_accolades[i] ) );
     }

     return SQLITE_OK;
}

int save_team( sqlite3 *db, const team_s *team )
{
     int rc;

     TRY( save_team_accolades(     db, team->accolades     ) );
     TRY( save_team_stats(         db, team->stats         ) );
     TRY( save_team_offense_stats( db, team->offense_stats ) );
     TRY( save_team_defense_stats( db, team->defense_stats ) );
     TRY( save_team_kicking_stats( db, team->kicking_stats ) );

     return SQLITE_OK;
}


static team_player_s *get_team_players_for_season( sqlite3 *db, const int team_id, const int season )
{
     static team_player_s sentinel = TEAM_PLAYER_SENTINEL;

     data_list_s list = { 0 };

     team_player_s team_player = { 0 };

     team_player.team_id = team_id;
     team_player.season  = season;

     if ( team_players_t_read_by_team_and_season( db, &team_player, &list ) != SQLITE_OK ) return NULL;

     if ( add_to_data_list( &list, &sentinel, sizeof(team_player_s), 10 ) < 0 )
     {
          free( list.data );

          return NULL;
     }

     return list.data;
}

static team_player_s *get_team_players( sqlite3 *db, const int team_id )
{
     static team_player_s sentinel = TEAM_PLAYER_SENTINEL;

     data_list_s list = { 0 };

     if ( team_players_t_read_by_team( db, team_id, &list ) != SQLITE_OK ) return NULL;

     if ( add_to_data_list( &list, &sentinel, sizeof(team_player_s), 10 ) < 0 )
     {
          free( list.data );

          return NULL;
     }

     return list.data;
}

static team_s *get_team_details( sqlite3 *db, const int team_id )
{
     team_s *team = NULL;

     if ( (team = malloc( sizeof(team_s) )) == NULL ) return NULL;

     memset( team, '\0', sizeof(team_s) );

     team->team_id = team_id;

     if ( teams_t_read( db, team ) != SQLITE_OK )
     {
          free( team );

          return NULL;
     }

     return team;
}

team_s *get_team_for_season( sqlite3 *db, const int team_id, const int season )
{
     team_s *team = NULL;

     if ( (team = get_team_details( db, team_id )) == NULL ) return NULL;

     team->players = get_team_players_for_season( db, team_id, season );

     return team;
}

team_s *get_team( sqlite3 *db, const int team_id )
{
     team_s *team = NULL;

     if ( (team = get_team_details( db, team_id )) == NULL ) return NULL;

     team->players = get_team_players( db, team_id );

     return team;
}


void free_team_players( team_player_s *team_players )
{
     for ( int i = 0; team_players[i].team_id > 0; ++i )
     {
          free_player( team_players[i].player );
     }

     free( team_players );
}

void free_team( team_s *team )
{
     if ( team == NULL ) return;

     if ( team->players ) free_team_players( team->players );

     free( team );
}

File: Ruby/team_service.rb
require 'team'
require 'team_accolade'
require 'team_accolades'
require 'team_player'
require 'team_stats'
require 'team_offense_stats'
require 'team_defense_stats'
require 'team_kicking_stats'

class TeamService

  def initialize repository, player_service
    @repository     = repository
    @player_service = player_service
  end

  def get_teams
    team = Team.new

    results = @repository.custom_read team.build_select_all

    teams = []

    unless results.nil? or results.length == 0
      results.each do |result|
        team.from_hash result

        teams.push team

        team = Team.new
      end
    end

    return teams
  end

  def get team_id
    team = @repository.read Team.new team_id

    unless team.nil?
      get_accolades     team
      get_players       team
      get_stats         team
      get_offense_stats team
      get_defense_stats team
      get_kicking_stats team
    end

    team
  end

  def get_accolades team
    team_accolade = TeamAccolade.new

    results = @repository.custom_read team_accolade.build_select_by_team, team.team_id

    unless results.nil? or results.length == 0
      team.accolades = []

      results.each do |result|
        team_accolade.from_hash result

        team.accolades.push team_accolade

        team_accolade = TeamAccolade.new
      end
    end
  end

  def get_players team
    team_players = TeamPlayer.new

    results = @repository.custom_read team_players.build_select_by_team, team.team_id

    unless results.nil? or results.length == 0
      team.players = []

      results.each do |result|
        team_players.from_hash result

        unless team_players.player_id.nil?
          team_players.player = @player_service.get team_players.player_id
        end

        team.players.push team_players

        team_players = TeamPlayer.new
      end
    end
  end

  def get_stats team
    team_stats = TeamStats.new

    results = @repository.custom_read team_stats.build_select_by_team, team.team_id

    unless results.nil? or results.length == 0
      team.stats = []

      results.each do |result|
        team_stats.from_hash result

        team.stats.push team_stats

        team_stats = TeamStats.new
      end
    end
  end

  def get_offense_stats team
    team_offense_stats = TeamOffenseStats.new

    results = @repository.custom_read team_offense_stats.build_select_by_team, team.team_id

    unless results.nil? or results.length == 0
      team.offense_stats = []

      results.each do |result|
        team_offense_stats.from_hash result

        team.offense_stats.push team_offense_stats

        team_offense_stats = TeamOffenseStats.new
      end
    end
  end

  def get_defense_stats team
    team_defense_stats = TeamDefenseStats.new

    results = @repository.custom_read team_defense_stats.build_select_by_team, team.team_id

    unless results.nil? or results.length == 0
      team.defense_stats = []

      results.each do |result|
        team_defense_stats.from_hash result

        team.defense_stats.push team_defense_stats

        team_defense_stats = TeamDefenseStats.new
      end
    end
  end

  def get_kicking_stats team
    team_kicking_stats = TeamKickingStats.new

    results = @repository.custom_read team_kicking_stats.build_select_by_team, team.team_id

    unless results.nil? or results.length == 0
      team.kicking_stats = []

      results.each do |result|
        team_kicking_stats.from_hash result

        team.kicking_stats.push team_kicking_stats

        team_kicking_stats = TeamKickingStats.new
      end
    end
  end

  def save team
    save_accolades     team.accolades
    save_players       team.players
    save_team_stats    team.stats
    save_offense_stats team.offense_stats
    save_defense_stats team.defense_stats
    save_kicking_stats team.kicking_stats

    @repository.create team
  end

  def save_accolades accolades
    return if accolades.nil?

    if accolades.is_a? Array
      accolades.each do |accolade|
        @repository.create accolade
      end
    else
      @repository.create accolades
    end
  end

  def save_players players
    return if players.nil?

    if players.is_a? Array
      players.each do |player|
        save_player player
      end
    else
      save_player players
    end
  end

  def save_player player
    return if player.nil?

    @repository.create player

    unless player.player.nil?
      @player_service.save player.player
    end
  end

  def save_team_stats team_stats
    return if team_stats.nil?

    if team_stats.is_a? Array
      team_stats.each do |stats|
        @repository.create stats
      end
    else
      @repository.create team_stats
    end
  end

  def save_offense_stats offense_stats
    return if offense_stats.nil?

    if offense_stats.is_a? Array
      offense_stats.each do |stats|
        @repository.create stats
      end
    else
      @repository.create offense_stats
    end
  end

  def save_defense_stats defense_stats
    return if defense_stats.nil?

    if defense_stats.is_a? Array
      defense_stats.each do |stats|
        @repository.create stats
      end
    else
      @repository.create defense_stats
    end
  end

  def save_kicking_stats kicking_stats
    return if kicking_stats.nil?

    if kicking_stats.is_a? Array
      kicking_stats.each do |stats|
        @repository.create stats
      end
    else
      @repository.create kicking_stats
    end
  end
end
